### ISSUE TITLE
Move managed cluster support (GKE) to general availability

### ISSUE THREAD

**Issue by salasberryfin** [kubernetes-sigs/cluster-api-provider-gcp#1168]
/kind feature

**Describe the solution you'd like**

As of today, support for provisioning GKE clusters is still an [experimental](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/tree/main/exp) feature and has been for a number of months.

In recent community meetings we discussed the viability of moving GKE support out of the experimental folder and into general availability. This would allow CAPG to follow other providers, such as CAPA and CAPZ which have full support for their corresponding managed clusters: [CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controllers/awsmanagedcluster_controller.go), [CAPZ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azuremanagedcluster_controller.go).

The promotion of GKE to stable could be a major step towards a more feature rich and stable CAPG, and we could push to add support for existing GKE functionalities that are still not covered by the provider (there are a few related PRs).

Since this is an important change, we thought it would be great to create this issue where we can discuss pros/cons and plan what needs to be done if we decide to go ahead.

Initially, this seems to be a non-breaking change for the API, since it means adding types but not modifying existing ones, but it would be great to have feedback from those that are more knowledgeable of the codebase. 

**Anything else you would like to add:**

As a follow up to this, `ClusterClass` support for managed clusters would be one of the relevant features we could be adding, as non-managed GCP clusters can already be provisioned from a class (refer to [this](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/402)) and this is expected to become a powerful/distinctive functionality of CAPI.

Please, feel free to share your thoughts on this.

### Related issues

- [x] https://github.com/kubernetes-sigs/cluster-api-provider-gcp/issues/1387
- [x] https://github.com/kubernetes-sigs/cluster-api-provider-gcp/issues/1309

**Comment by damdo** [comment]
@salasberryfin is looking for some feedback on this one from core maintainers
/assign @richardcase @cpanato

**Comment by salasberryfin** [comment]
Looking at unblocking this and moving the feature out of experimental.

Let's collect a list of PRs related to managed clusters that we should address:
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/1024
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/901
- [ ] 

Feel free to contribute to the list with other relevant PRs/issues.

cc: @richardcase @damdo @cpanato

**Comment by jasonumiker** [comment]
As far as I can tell there is no way today to ask for a Zonal rather than a regional GKE today. Given that GCP gives you one of those for free this is an important feature for many in their dev/lab environments that I'd hope to see in GA.

**Comment by abasu0713** [comment]
Hey folks.. Just following up on this.. My org needs the ability to manage fleets across multiple projects and we are conceptualizing a CAPG managed GKE solution.. We definitely need BinAuthz, Dataplane V2, and Gateway API enablement. The project timeline is flexible. We would really appreciate if ya'll can provide us an update please

**Comment by jonathanrainer** [comment]
# Progress Report

This issue has been open over 2 years now so it's worth a moment to take stock of what has been implemented and what there is still to do. So this comment will act as an audit of current work and an action plan to move us towards GA in the near future.

Conducting this audit was performed by looking at existing features that are either labelled for GKE or touch specifically GKE objects (i.e. `GCPManagedControlPlane` etc.). Working out the remaining gaps was done by sweeping the repo for GKE related issues and feature requests and correlating that with other more umbrella issues like #1024 and others mentioned within this issue itself. In addition I compared what CAPG offers with CAPA and CAPZ (both of which have managed offerings in GA) to spot any other obvious gaps and that produced the list(s) below.

## What Has Been Done?
Lots of GKE features have already been shipped and confirmed as working in production environments, including:
* Private Clusters - #1186 
* Master-authorized-networks - #1004 
* Pod-level Workload Identity - #1504 
* Binary Authorisation - #1504 
* Release Channel Support - #770 & #1498
* Logging & Monitoring Service Selection - #1292 
* Shared VPC Networking - #1189
* Regional & Zonal Clusters - #764, #770, #1289
* GKE ClusterClasses - #1442 
* Nodepool Autoscaling - #789 
* Node-pool auto-upgrade & auto-repair - #1026 
* Node-level shielded config & GKE Sandbox/gVisor - #944 
* Pre-emptible & Spot VMs - #1659 
* Autopilot Support - #764 (extended later with other PRs)
* Gateway API Support - #1733 
* Controller WIF Support - #1639 

This has been a monumental effort and we have many authors to thank for getting to this point. However, by comparison to particularly CAPA and CAPZ this isn't enough to declare GKE as GA, there are several important features that are either in-progress or still need to be done in order to reach that point.

# What Is In Progress?

Closing the gap between what we have now and GA is happening, with many features currently in PR that would address some of the functionality gaps and bugs that currently exist in the implementations:

* Secret Manager CSI & Secret Sync - #1741 
* Support for secondary range names - #1637 
* Adding label propagation from CRs to GKE Objects - #1646
* Enabling GKE E2E Tests - #1759 
    * This is dependent on 3 other PRs, due to some foundational work being required before the tests can be enabled:
      * Enable scaling correctly under GKE - #1657 
      * Add a nil-guard to Gateway API channel checks - #1758 
      * Change how we mint tokens for KubeConfig to remove need for IAM changes for E2Es - #1750 
* Maintenance Policies - #1751 
* DatapathProvider & DNS Config - #1752 
* AddonsConfig - #1753 
* Fixing Network and Subnet Immutability - #1676 (Comments on PR note this should address GKE too), #1754 (more specific GKE cases)

## What is still to be done?

So as can be seen there is a large amount of work in-flight to enable many of the features that would reach parity with the offerings in CAPA and CAPZ, however there are still some that are outstanding:

* Add support for a few extra fields (can probably be a single PR):
    * `ShieldedNodes`
    * `DefaultMaxPodsConstraint` and `ClusterIpv4Cidr`
    * `PrivateEndpointSubnetwork`
    * `LoggingConfig` as a replacement for `LoggingService` though it may well live alongside it for now
* Fix up `BinaryAuthorization` support so it can be updated and not just set at creation time (the GKE API allows this)
* `ConfidentialCompute` wiring for GKE node pools (already exists for bare Compute Engine VMs)
* Cluster level autoscaling/Node auto-provisioning
* Vertical Pod Autoscaling Toggle
* Fix support for `Cluster.NetworkPolicy`, this is the legacy alternative to the Dataplane stuff but we may as well fully support both, especially because PR #1133 has done most of the work it just needs reviving and reworking for the new code base.

## Action Plan
So with all that in mind where does that leave us in terms of GKE support. I'd estimate we're about 50 - 60% of the way there, with the items listed above forming the basis for the feature set that GA would deliver. To get us the rest of the way there we should do the following:

1. Get all in progress items listed above reviewed and merged
    * We should focus on those that block GKE E2E tests in the first instance and then rebase other PRs accordingly as this is one of the best signals that we have that the new features will work.
2. Raise issues/PRs to cover the remaining items that are still to be done and get those reviewed and merged
3. Move GKE types out of `exp` and into a stable API Package 
    * So as to make life less complicated this should happen after the move to v1beta2 types that Carlos is currently working on (#1607), so the GKE types become part of v1beta2 as an addition
4. Conduct a documentation audit and update so as to make sure all the new features are well documented, most of the in progress PRs have this but the majority of the historic ones do not. We should find the gaps and close them.
4. Remove the feature gate, turning on GKE support by default

I'm happy to hear arguments about cutting scope, especially for the work that is not currently in-flight  but if we're happy with this action plan I can start raising issues for the remaining gaps and then we can start to pick them off one at a time.

In summary, we've done a lot of work to add features and stabilise GA and have a lot of work in flight to really broaden out the feature surface. What remains is quite small and tactical (beyond lots of reviews) and as such I don't believe that GA is far away but we will need concerted effort to get it over the line.
